dali/python/nvidia/dali/pool.py

```diff
@@ -0,0 +1,327 @@
+        self.pool = pool
+        self.rec_pipes = self.pool.res_pipes + [self.pool.from_tracker]
+
+    @classmethod
+    def from_groups(cls, groups, workers_num, init_method, keep_alive_queue_size, initial_chunk_size=1024 * 1024):
+        """Creates new WorkerPool instance for given list of ExternalSource groups.
+
+        Parameters
+        ----------
+        `groups` : _ExternalSourceGroup list
+            List of external source groups.
+        `workers_num` : int
+            Number of workers to be created in ProcPool.
+        `init_method` : str
+            Method of starting worker processes, either fork or spawn.
+        `keep_alive_queue_size` : int
+            Number of most recently produced batches that should be kept alive and remain untouched (because they might be
+            still referenced further in the pipeline). Note that actual number of simultaneously kept batches
+            will probably be grater because of additional prefetching.
+        `initial_chunk_size` : int
+            Initial size of each shared memory chunk.
+        """
+        callbacks = [group.callback for group in groups]
+        queue_depths = [keep_alive_queue_size + group.prefetch_queue_depth for group in groups]
+        pool = ProcPool(callbacks, workers_num, init_method, queue_depths, initial_chunk_size)
+        return cls(len(callbacks), pool)
+
+    def schedule_batch(self, context_i, batch_i, tasks):
+        """Distribute `tasks` among workers to run them by calling `context_i`th callaback
+
+        Parameters
+        ----------
+        `context_i` : int
+            Specifies which callback will be used to run the task, it must be the index corresponding
+            to the order of callbacks passed when constructing WorkerPool.
+        `batch_i` : int
+            Ordinal of the batch that tasks list corresponds to.
+        `tasks` : nvidia.dali.types.SampleInfo list
+            You can think of resulting batch as [callback(task) for task in tasks] with the exception that
+            callbacks will be run in parallel.
+        """
+        tasks = list(enumerate(tasks))
+        if not tasks:
+            raise RuntimeError("Cannot schedule empty list of tasks")
+        context = self.contexts[context_i]
+        if batch_i in context.scheduled:
+            raise RuntimeError("Given batch has already been scheduled")
+        self._distribute((context_i, batch_i), tasks)
+        context.partially_received[batch_i] = {}
+        context.scheduled[batch_i] = tasks
+
+    def _distribute(self, header, tasks):
+        workers_num = self.pool.workers_num
+        tasks_no = len(tasks)
+        chunk_size = tasks_no // workers_num
+        queued_no = chunk_size + (tasks_no % workers_num)
+        with self.pool.task_pipes_lock:
+            task_pipes = self.pool.task_pipes
+            task_pipes[0].send((header, tasks[:queued_no]))
+            if chunk_size:
```

Comment:
What if tasks_no < workers_num? chunk_size will be 0. I think you need to handle the tail as chunk_size is rounded down.

Reply:
The ``if chunk_size`` was exactly for that, just to pass the tail to the first worker in such case. Do you suggest spreading the tail one - by - one to the first tasks_no workers?

Reply:
Now I get it. 
queued_no is the tail, right? So why it is not:
```
queued_no  = tasks_no % workers_num
```

Reply:
NVM. You send chunk_size + tail to the first worker.
Maybe even distribution would be better anyway.

Reply:
changed to distributing remaining tasks equally across first len(remaining) workers.